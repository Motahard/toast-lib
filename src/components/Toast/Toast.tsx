import React, { MouseEvent, useEffect, useState } from 'react'
import { ThemeProvider } from 'styled-components'

import { instanceToast } from '@src/logic/ToastService'
import { 
  Delays,
  Animations,
  Positions
} from '@src/constants'
import { changeAnimation } from '@src/helpers/utils'
import { ICreateToast } from '@src/interfaces'

import { defaultTheme } from '@src/styles/theme'
import {
  ToastWrapper,
  Logo,
  TextContainer,
  Title,
  Description,
  CloseButton
} from '@src/components/Toast/styled'

export const Toast: React.FC<ICreateToast> = ({
    id, 
    type,
    title, 
    description, 
    autoDelete = false, 
    delayForDelete = Delays.DEFAULT_DELAY_DELETE,
    position, 
    animation,
    backgroundColor,
    textColor,
    story = false
  }) => {
    const [currentAnimation, setCurrentAnimation] = useState(animation)
    const [show, setShow] = useState(true)
    let positionX: number
    let positionY: number

    useEffect(() => {
      if(autoDelete) {
        if(story) {
          setTimeout(() => setShow(false), delayForDelete)
        } else {
          instanceToast.removeToastWithDelay(id, delayForDelete)
        }
        
        const changeAnimTimeout = 
        setTimeout(() => {
          setCurrentAnimation(changeAnimation(currentAnimation))
        }, delayForDelete - Delays.DEFAULT_ANIM_DELAY)

        return () => clearTimeout(changeAnimTimeout)
      }
    }, [])
    
    const handleDelete = () => {
      setCurrentAnimation(changeAnimation(currentAnimation))
      setTimeout(() => {
        if(story) setShow(false)
        instanceToast.removeToast(id)
      }, Delays.DEFAULT_ANIM_DELAY)
    }

    const handleMouseDown = (e: MouseEvent<HTMLDivElement>) => {
      positionX = e.clientX
      positionY = e.clientY
    }

    const handleMouseUp = (e: MouseEvent<HTMLDivElement>) => {
      if (position === Positions.LEFT_TOP || position === Positions.LEFT_BOTTOM) {
        switch (animation) {
          case Animations.FROM_LEFT: {
            positionX > e.clientX && handleDelete()
            break
          }
          case Animations.FROM_TOP: {
            positionY > e.clientY && handleDelete()
            break
          }
          case Animations.FROM_BOTTOM: {
            positionY < e.clientY && handleDelete()
            break
          }
          default:
            break
        }
      }
      else if (position === Positions.RIGHT_TOP || position === Positions.RIGHT_BOTTOM) {
        switch (animation) {
          case Animations.FROM_RIGHT: {
            positionX < e.clientX && handleDelete()
            break
          }
          case Animations.FROM_TOP: {
            positionY > e.clientY && handleDelete()
            break
          }
          case Animations.FROM_BOTTOM: {
            positionY < e.clientY && handleDelete()
            break
          }
          default:
            break
        }
      }
    }

    if(!show) {
      return null
    }
    return (
      <ThemeProvider theme={defaultTheme}>
        <ToastWrapper 
          type={type} 
          backgroundColor={backgroundColor}
          position={position} 
          animation={currentAnimation}
          onMouseUp={handleMouseUp}
				  onMouseDown={handleMouseDown}
        >
          <Logo type={type}/>
          <TextContainer type={type} textColor={textColor}>
            <Title>
              {title}
            </Title>
            {description && (
              <Description>
                {description}
              </Description>
            )}
          </TextContainer>
          <CloseButton onClick={handleDelete} type={type} />
        </ToastWrapper>
      </ThemeProvider>
  )
}