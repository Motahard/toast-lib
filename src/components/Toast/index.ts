export * from '@components/Toast/Toast'