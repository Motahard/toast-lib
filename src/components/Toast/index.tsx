import React, { DragEvent, MouseEvent, useEffect, useRef, useState } from 'react'
import { ThemeProvider } from 'styled-components'

import { instanceToast } from '@services'
import { 
  Delays,
  Animations
} from '@constants'
import { changeAnimation } from '@utils'
import { ICreateToast } from '@interfaces'

import { defaultTheme } from '@styles'
import {
  ToastWrapper,
  Logo,
  TextContainer,
  Title,
  Description,
  CloseButton
} from '@components/Toast/styled'

export const Toast: React.FC<ICreateToast> = ({
    id, 
    type,
    title, 
    description, 
    autoDelete = false, 
    delayForDelete = Delays.DEFAULT_DELAY_DELETE,
    position, 
    animation,
    backgroundColor,
    textColor,
    story = false
  }) => {
    const [currentAnimation, setCurrentAnimation] = useState(animation)
    const [show, setShow] = useState(true)
    const changeAnimTimeout: { current: NodeJS.Timeout | null } = useRef(null)

    let [positionX, setPositionX] = useState<number>()

    useEffect(() => {
      if(autoDelete) {
        if(story) {
          setTimeout(() => setShow(false), delayForDelete)
        } else {
          instanceToast.removeToastWithDelay(id, delayForDelete)
        }
        
        changeAnimTimeout.current = 
        setTimeout(() => {
          setCurrentAnimation(changeAnimation(currentAnimation))
        }, delayForDelete - Delays.DEFAULT_ANIM_DELAY)
      }
    }, [])
    
    const handleDeleteClick = (e: MouseEvent) => {
      setCurrentAnimation(changeAnimation(currentAnimation))
      setTimeout(() => {
        if(story) setShow(false)
        instanceToast.removeToast(id)
      }, Delays.DEFAULT_ANIM_DELAY)
    }

    const handleDelete = () => {
      setTimeout(() => {
        if(story) setShow(false)
        instanceToast.removeToast(id)
      }, Delays.DEFAULT_ANIM_DELAY)
    }

    const onMouseDown = (e: DragEvent<HTMLDivElement>) => {
      e.preventDefault()
      setPositionX(e.clientX)
    }

    const onMouseMove = (e: MouseEvent<HTMLDivElement>) => {
      if(!positionX || (positionX === e.clientX)) return
      if(positionX > e.clientX) {
        setCurrentAnimation(Animations.TO_LEFT)
        handleDelete()
      } else if (positionX < e.clientX) {
        setCurrentAnimation(Animations.TO_RIGHT)
        handleDelete()
      }
    }

    if(!show) {
      return null
    }
    return (
      <ThemeProvider theme={defaultTheme}>
        <ToastWrapper 
          draggable
          type={type} 
          backgroundColor={backgroundColor}
          position={position} 
          animation={currentAnimation}
          onMouseDown={onMouseDown}
          onMouseUp={onMouseMove}
        >
          <Logo type={type}/>
          <TextContainer type={type} textColor={textColor}>
            <Title>
              {title}
            </Title>
            {description && (
              <Description>
                {description}
              </Description>
            )}
          </TextContainer>
          <CloseButton onClick={handleDeleteClick} type={type} />
        </ToastWrapper>
      </ThemeProvider>
  )
}