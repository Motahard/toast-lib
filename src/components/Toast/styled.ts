import styled from "styled-components";
import { Types, Delays } from '@src/constants'
import { 
	font,	
	fromBottom,
	fromLeft,
	fromRight,
	fromTop,
	toBottom,
	toLeft,
	toRight,
	toTop,
	svgObj
} from '@src/styles/theme'

import { 
	IStyledToastWrapper, 
	IStyledToast, 
	IStyledToastText 
} from "@src/interfaces";

export const ToastWrapper = styled.div<IStyledToastWrapper>`
    position: relative;
	display: flex;
	align-items: center;
	margin-bottom: 10px;
	width: 300px;
	height: 100px;
	border-radius: 24px;
	box-shadow: 4px 4px 8px #00000029;
	opacity: 1;
	padding: 16px;
	transition: transform ${Delays.DEFAULT_ANIM_DELAY}ms ease-in;
	cursor: grab;
    background: ${(props) => {
		if(props.backgroundColor) {
			return props.backgroundColor 
		} else {
			switch (props.type) {
				case Types.INFO:
					return props.theme.type.info
				case Types.WARNING:
					return props.theme.type.warning
				case Types.ERROR:
					return props.theme.type.error
				default:
					return props.theme.type.success
			}
		}
	}};
    ${font}
	animation: ${({ animation }) => animation} 1s;
	${fromLeft};
	${toLeft};
	${fromTop};
	${toTop};
	${fromRight};
	${toRight};
	${fromBottom};
	${toBottom};
	&:active {
		cursor: grabbing;
	}
`
export const Logo = styled.div<IStyledToast>`
    height: 64px;
	width: 64px;
	${(props) => {
		switch (props.type) {
			case Types.INFO:
				return `background-image: url(${svgObj.info})`
			case Types.WARNING:
				return `background-image: url(${svgObj.warning})`
			case Types.ERROR:
				return `background-image: url(${svgObj.error})`
			default:
				return `background-image: url(${svgObj.success})`
		}
	}};
`
export const TextContainer = styled.div<IStyledToastText>`
	height: auto;
	margin-left: 16px;
	color: ${(props) => {
		if(props.textColor) {
			return props.textColor
		} else {
			return props.type === Types.WARNING ? 
			props.theme.textColor.black
			: 
			props.theme.textColor.white 
		}
	}};
`
export const Title = styled.h2`
    font-size: 28px;
	margin: 0;
`
export const Description = styled.p`
	font-size: 18px;
    margin: 0;
`
export const CloseButton = styled.div<IStyledToast>`
    cursor: pointer;
    position: absolute;
	height: 32px;
	width: 32px;
	right: 29px;
	top: 27px;
	${(props) =>
		props.type === Types.WARNING
			? `background-image: url(${svgObj.closeBlack});`
			: `background-image: url(${svgObj.closeWhite});`}
`