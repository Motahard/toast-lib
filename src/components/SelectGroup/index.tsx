import React, { FC } from 'react'
import { v4 as uuid } from 'uuid'
import { ISelectGroup } from '@interfaces'
import { SelectContainer, SelectTitle, Select, SelectOption } from '@components/SelectGroup/styled'

export const SelectGroup: FC<ISelectGroup> = ({ name, values, onChange, description, valuesDescription, value }) => {
  return (
    <SelectContainer>
        <SelectTitle htmlFor={name}>{description}</SelectTitle>
        <Select 
            value={value}
            name={name}
            id={name}
            onChange={onChange}
        >
            {values.map((item, index) => <SelectOption key={uuid()} value={item}>{valuesDescription[index]}</SelectOption>)}
        </Select>
    </SelectContainer>
  )
}