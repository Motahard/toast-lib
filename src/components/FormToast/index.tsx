import React, { ChangeEvent, Component, FormEvent } from 'react'
import { Delays } from '@constants'
import { instanceToast } from '@services'
import { IFormToastProps, IFormToastState } from '@interfaces'
import { InputGroup } from '@components/InputGroup'
import { SelectGroup } from '@components/SelectGroup'
import { Form, Button } from '@components/FormToast/styled'

export default class FormToast extends Component<IFormToastProps, IFormToastState> {  
    constructor(props: IFormToastProps) {
        super(props)
        this.state = {
            type: 'success',
            title: '',
            description: '',
            backgroundColor: '',
            textColor: '',
            autoDelete: true,
            delayForDelete: Delays.DEFAULT_DELAY_DELETE
        }
    }

    handleChange = (e: ChangeEvent<HTMLInputElement | HTMLSelectElement>) => {
        const { name, value } = e.target
        if (name === 'autoDelete') {
            this.setState((prev) => ({autoDelete: !prev.autoDelete}))
        } else {
            this.setState({
                [name]: value
            } as Pick<IFormToastState, any>)
        }
    }

    handleSubmit = (e: FormEvent<HTMLFormElement>) => {
        e.preventDefault()
        instanceToast.generateToast({...this.state})
    }
    
    render() {
        return (
            <Form onSubmit={this.handleSubmit}>
                <SelectGroup 
                    name='type' 
                    values={['success', 'info', 'warning', 'error']}
                    valuesDescription={['Success', 'Information', 'Warning', 'Error']}
                    onChange={this.handleChange}
                    description='Type Toast*'
                    value={this.state.type}
                />
                <InputGroup 
                    type='text' 
                    name='title' 
                    value={this.state.title} 
                    onChange={this.handleChange} 
                    description='Toast Title*'
                    placeholder='Enter toast title'
                />
                <InputGroup 
                    type='text' 
                    name='description' 
                    value={this.state.description} 
                    onChange={this.handleChange} 
                    description='Toast Description (optional)'
                    placeholder='Enter toast description'
                />
                <InputGroup 
                    type='text' 
                    name='backgroundColor' 
                    value={this.state.backgroundColor} 
                    onChange={this.handleChange} 
                    description='Background Color (optional)'
                    placeholder='Enter toast background color'
                />
                <InputGroup 
                    type='text' 
                    name='textColor' 
                    value={this.state.textColor} 
                    onChange={this.handleChange} 
                    description='Text Color (optional)'
                    placeholder='Enter toast text color'
                />
                <InputGroup 
                    type='checkbox' 
                    name='autoDelete' 
                    value={(this.state.autoDelete)?.toString()} 
                    onChange={this.handleChange} 
                    description='No autodelete'
                />
                <InputGroup 
                    type='text' 
                    name='delayForDelete' 
                    value={this.state.delayForDelete} 
                    onChange={this.handleChange} 
                    description='Delay For Delete'
                    placeholder='Enter toast delay for delete'
                />
                <Button type='submit'>
                     Add
                </Button>
            </Form>
    )
  }
}
