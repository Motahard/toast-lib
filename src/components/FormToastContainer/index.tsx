import React, { ChangeEvent, FC } from 'react'
import { IFormToastContainer } from '@interfaces'
import { 
    selectDescriptions, 
    selectDescripValues, 
    selectNames, 
    selectValues 
} from '@constants'
import { InputGroup } from '@components/InputGroup'
import { SelectGroup } from '@components/SelectGroup'
import { Form } from '@components/FormToast/styled'

const FormToastContainer: FC<IFormToastContainer> = ({ setAnimation, setPosition, setSpace, space }) => {
    const handlePositionChange = (e: ChangeEvent<HTMLSelectElement>) => {
        const { value } = e.target
        setPosition(value)
    }

    const handleAnimationChange = (e: ChangeEvent<HTMLSelectElement>) => {
        const { value } = e.target
        setAnimation(value)
    }

    const handleSpaceChange = (e: ChangeEvent<HTMLInputElement>) => {
        const { value } = e.target
        setSpace(value)
    }

    const handlers = [handlePositionChange, handleAnimationChange]
    return (
        <Form>
            {selectNames.map((item, index) => (
                <SelectGroup 
                    key={item}
                    name={item}
                    values={selectValues[index]}
                    valuesDescription={selectDescripValues[index]}
                    onChange={handlers[index]}
                    description={selectDescriptions[index]}
                />
            ))}
            <InputGroup 
                type='text' 
                name='space' 
                value={space} 
                onChange={handleSpaceChange} 
                description='Toast Space (px)' 
                placeholder='Enter toast space in px'
            />
        </Form>
  )
}

export default FormToastContainer