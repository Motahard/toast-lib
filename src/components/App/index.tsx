import React, { useState } from 'react';
import FormToast from '@components/FormToast';
import FormToastContainer from '@components/FormToastContainer';
import { ToastInit } from '@components/ToastInit';

export const App = () => {
  const [position, setPosition] = useState('RIGHT_TOP')
  const [animation, setAnimation] = useState('FROM_RIGHT')
  const [space, setSpace] = useState('0');

  return <>
    <FormToastContainer 
      setPosition={setPosition} 
      setAnimation={setAnimation} 
      setSpace={setSpace} 
      space={space}
    />
    <FormToast />
    <ToastInit 
      position={position} 
      animation={animation} 
      space={space}
    />
  </>;
}
