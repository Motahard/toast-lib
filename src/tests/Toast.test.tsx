import React from 'react';
import renderer from 'react-test-renderer';
import { Toast } from '../components/Toast';

it('rendering toast', () => {
  const component = renderer.create(
    <Toast 
      id='1'
      type='success'
      title='hi'
      description='hello'
      position='LEFT_TOP'
      animation='FROM_LEFT'
      space='0'
    />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});