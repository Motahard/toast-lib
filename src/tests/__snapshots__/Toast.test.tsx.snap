// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering toast 1`] = `
<div
  className="sc-bcXHqe cBfWNP"
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  type="success"
>
  <div
    className="sc-gswNZR bblguL"
    type="success"
  />
  <div
    className="sc-dkrFOg uzKwX"
    type="success"
  >
    <h2
      className="sc-hLBbgP hNGiGF"
    >
      hi
    </h2>
    <p
      className="sc-eDvSVe kxdvCy"
    >
      hello
    </p>
  </div>
  <div
    className="sc-jSUZER VCAhw"
    onClick={[Function]}
    type="success"
  />
</div>
`;
